### ISSUE TITLE
What features would you like to see in FastMoE v0.2?

### ISSUE THREAD

**Issue by laekov** [laekov/fastmoe#11]
Open for suggestions and discussions

**Comment by laekov** [comment]
* expose gate score & idx to users
* fp16 backward time inspection
* support zero optimizer
* model load / save

**Comment by wodsoe** [comment]
Could you write an example for stylegan2-ada, https://github.com/NVlabs/stylegan2-ada-pytorch

**Comment by xptree** [comment]
> Could you write an example for stylegan2-ada, https://github.com/NVlabs/stylegan2-ada-pytorch

Could you be more clear? According to the stylegan2-ada paper (https://arxiv.org/pdf/2006.06676.pdf), its main contribution is "an adaptive discriminator augmentation mechanism" which "does not require changes to loss functions or network architectures".  If I understand correctly, the augmentation mechanism is orthogonal to the choice of detailed model architectures, e.g., whether to use MoE or not.

**Comment by xptree** [comment]
@laekov We should also change the API of MoELinear to accept bias. I.e., from `MOELinear.apply(input, weight, expert_count)` to `MOELinear.apply(input, weight, bias, expert_count)`, where `bias` could be `None`.

**Comment by lisuq** [comment]
Could you write an example for transformers?

**Comment by laekov** [comment]
> Could you write an example for transformers?

We do have 2 examples, Megatron-LM and Transformer-XL, which are both popular transformer models, and the examples show how to use FastMoE using different levels of APIs. Can you please explain in detail that what level of example would you like to see? 

BTW, contributions are highly appreciated. You can contribute to our example codebase if you are interested.

**Comment by Nathan-For** [comment]
Where to download the parameters？

**Comment by laekov** [comment]
> Where to download the parameters？

FastMoE is a system for training MoE models. We do not have any product model yet.

**Comment by Crishawy** [comment]
Can MoE be applied to CNN kind network?

**Comment by laekov** [comment]
> Can MoE be applied to CNN kind network?

Yes. You can use any CNN layer as the expert.
